File: Models/Models/DTO/Agent.cs
using Models.DB;
using Models.Enums;
using System.Net.NetworkInformation;

namespace Models.DTO
{
    public class PeopleHeirarchyDto
    {
        public int? AgentId { get; set; }
        public long? HierarchyId { get; set; }
        public string? AgentCode { get; set; } = null!;
        public string? FirstName { get; set; }
        public string? MiddleName { get; set; }
        public string? LastName { get; set; }
        public PeopleHeirarchyDto? Supervisors { get; set; }
        public string? HierarchyPath { get; set; } = string.Empty;
    }
    public class AgentDto
    {
        public int AgentId { get; set; }
        public string AgentCode { get; set; } = null!;
        public string? AgentTypeCode { get; set; }
        public string? AgentSubTypeCode { get; set; }
        public string? AgentName { get; set; }
        public string? BusinessName { get; set; }
        public string? FirstName { get; set; }
        public string? MiddleName { get; set; }
        public string? LastName { get; set; }
        public string? Prefix { get; set; }
        public string? Suffix { get; set; }
        public string? Gender { get; set; }
        public DateTime? DOB { get; set; }
        public string? Nationality { get; set; }
        public string? MaritalStatusCode { get; set; }
        public string? PreferredLanguage { get; set; }
        public string? ChannelCode { get; set; }
        public string? SubChannelCode { get; set; }
        public string? DesignationCode { get; set; }
        public string? Designation { get; set; }
        public string? AgentLevel { get; set; }
        public string? LocationCode { get; set; }
        public string? StaffCode { get; set; }
        public int? Supervisor_Id { get; set; }
        public DateTime? ContractedDate { get; set; }
        public string? AgentStatusCode { get; set; }
        public DateTime? StatusDate { get; set; }
        public bool IsLicensed { get; set; }
        public string? MaskedPanNumber { get; set; }
        public string? aadhaar_number { get; set; }
        public string? IrdaLicenseNumber { get; set; }
        public string? GstNumber { get; set; }
        public string CreatedBy { get; set; } = null!;
        public DateTime CreatedDate { get; set; }
        public string? ModifiedBy { get; set; }
        public DateTime? ModifiedDate { get; set; }
        public int? RowVersion { get; set; }
        public bool IsActive { get; set; } = true;
        public List<AgentDto>? Supervisors { get; set; }
        public List<AgentDto>? Reportees { get; set; }
        public List<AgentAuditTrailDTO>? agentAuditTrail { get; set; }
        public List<PeopleHeirarchyDto>? peopleHeirarchy { get; set; }
        public string? CandidateType { get; set; }
        public string? ApplicationDocketNo { get; set; }
        public string? Title { get; set; }
        public string? Father_Husband_Nm { get; set; }
        public string? Channel_Name { get; set; }
        public string? Sub_Channel { get; set; }
        public string? EmployeeCode { get; set; }
        public DateTime? StartDate { get; set; }
        public bool PanAadharLinkFlag { get; set; }
        public bool Sec206abFlag { get; set; }
        public List<Nominee>? nominees { get; set; }
        public string? PackageID { get; set; }
        public List<PersonalInfo>? personalInfo { get; set; }
        public string? CommissionClass { get; set; }
        public string? TaxStatus { get; set; }
        public string? StateEid { get; set; }
        public int? OccupationCode { get; set; }
        public String? Occupation { get; set; }
        public string? URN { get; set; }
        public string? AdditionalComment { get; set; }
        public DateTime? AppointmentDate { get; set; }
        public DateTime? IncorporationDate { get; set; }
        public string? CnctPersonDesig { get; set; }
        public string? CnctPersonMobileNo { get; set; }
        public string? CnctPersonEmail { get; set; }
        public string? CnctPersonName { get; set; }
        public string? AgentTypeCategory { get; set; }
        public string? AgentClassification { get; set; }
        public string? CMSAgentType { get; set; }
        public List<BankAccount>? bankAccounts { get; set; }
        public string? ServiceTaxNo { get; set; }
        public List<Address>? PermanentAddres { get; set; }
        public List<Address>? MailingAddres { get; set; }
        public bool UlipFlag { get; set; } = false;
        public string? TrainingGroupType { get; set; }
        public string? Ifs { get; set; }
        public bool RefresherTrainingCompleted { get; set; }
        public bool IsMigrated { get; set; }
        public string? MainPartnerClientCode { get; set; }
        public string? AgentMaincodevwEid { get; set; }
        public DateTime? RegistrationDate { get; set; }
        public string? Vertical { get; set; }
        public string? BranchCode { get; set; }
        public string? BranchName { get; set; }
        public DateTime? Ic36TrngCompletionDate { get; set; }
        public DateTime? STrngCompletionDate { get; set; }
        public DateTime? ConfirmationDate { get; set; }
        public DateTime? FgRockstarTrainingDate { get; set; }
        public DateTime? IncrementDate { get; set; }
        public DateTime? LastPromotionDate { get; set; }
        public DateTime? HRDoj { get; set; }
        public DateTime? FgValueTrngDate { get; set; }
        public DateTime? HSecPolicyTrngDate { get; set; }
        public DateTime? ItSecPolicyTrngDate { get; set; }
        public DateTime? NpsTrngCompletionDate { get; set; }
        public DateTime? WhistleBlowerTrngDate { get; set; }
        public DateTime? GovPolicyTrngDate { get; set; }
        public DateTime? InductionTrngDate { get; set; }
        public DateTime? LastWorkingDate { get; set; }
        public string? LicenseNo { get; set; }
        public string? LicenseType { get; set; }
        public DateTime? LicenseIssueDate { get; set; }
        public DateTime? LicenseExpiryDate { get; set; }
        public string? LicenseStatus { get; set; }
        public List<KeyValueEntry>? bankAccType { get; set; }
        public List<KeyValueEntry>? titles { get; set; }
        public List<KeyValueEntry>? genders { get; set; }
        public List<KeyValueEntry>? channelNames { get; set; }
        public List<KeyValueEntry>? subChannels { get; set; }
        public List<KeyValueEntry>? occupations { get; set; }
        public List<KeyValueEntry>? agentTypeCategories { get; set; }
        public List<KeyValueEntry>? agentClassifications { get; set; }
        public List<KeyValueEntry>? maritalStatuses { get; set; }
        public List<KeyValueEntry>? educationCodes { get; set; }
        public List<KeyValueEntry>? stateNames { get; set; }
        public List<KeyValueEntry>? countries { get; set; }
    }
    public class AgentListRequest
    {
        public string userid
        {
            get; set;
        }
    }

    public class SearchAgent
    {
        public string? SearchCondition { get; set; }
        public string? Zone { get; set; }
        public int? AgentId { get; set; } = null!;
        public string? AgentCode { get; set; } = null!;
        //public string? AgentName { get; set; }
        //public string? ChannelCode { get; set; }
        //public string? SubChannelCode { get; set; }
        //public string? PanNumber { get; set; }
        //public string? Email { get; set; }
        //public string? MobileNo { get; set; }
        //public string? AadhaarNumber { get; set; }
        //public string? IrdaLicenseNumber { get; set; }
        //public string? GstNumber { get; set; }
        public Int64? PageNo { get; set; }
        public Int64? PageSize { get; set; }
        public string? SortColumn { get; set; }
        public string? SortDirection { get; set; }
        public bool FetchHierarchy { get; set; } = false;
    }
    public class AgentDtoResponse
    {
        public int agent_id { get; set; }
        public string agent_code { get; set; } = null!;
        public string? AgentTypeCode { get; set; }
        public string? AgentSubTypeCode { get; set; }
        public string? agent_name { get; set; }
        public string? BusinessName { get; set; }
        public string? FirstName { get; set; }
        public string? MiddleName { get; set; }
        public string? LastName { get; set; }
        public string? Prefix { get; set; }
        public string? Suffix { get; set; }
        public string? Gender { get; set; }
        public DateTime? DOB { get; set; }
        public string? Nationality { get; set; }
        public string? MaritalStatusCode { get; set; }
        public string? PreferredLanguage { get; set; }
        public string? ChannelCode { get; set; }
        public string? SubChannelCode { get; set; }
        public string? DesignationCode { get; set; }
        public string? AgentLevel { get; set; }
        public string? LocationCode { get; set; }
        public string? StaffCode { get; set; }
        public int? Supervisor_Id { get; set; }
        public DateTime? ContractedDate { get; set; }
        public string? AgentStatusCode { get; set; }
        public DateTime? StatusDate { get; set; }
        public bool IsLicensed { get; set; }
        public string? MaskedPanNumber { get; set; }
        public string? aadhaar_number { get; set; }
        public string? irda_license_number { get; set; }
        public string? gst_number { get; set; }
        public string CreatedBy { get; set; } = null!;
        public DateTime CreatedDate { get; set; }
        public string? ModifiedBy { get; set; }
        public DateTime? ModifiedDate { get; set; }
        public int? RowVersion { get; set; }
        public bool IsActive { get; set; } = true;
        public string? pan_number { get; set; }
        public string? Email { get; set; }
        public string? MobileNo { get; set; }
        public int? total_count { get; set; }

    }
}

File: UI/Admin-ui/src/models/agent.ts
export interface IAgentSearchRequest {
  searchCondition: string
  zone: string
}
export interface IAgentSearchByCodeRequest {
  searchCondition?: string
  zone?: string
  agentId?: number
  agentCode?: string
  pageNo?: number
  pageSize?: number
  sortColumn?: string
  sortDirection?: string
  FetchHierarchy?: boolean
}

export interface IAgent {
  agentId: number
  agentCode: string
  agentTypeCode: string | null
  agentSubTypeCode: string | null
  agentName: string
  businessName: string | null
  maritalStatusCode: string

  firstName: string | null
  middleName: string | null
  lastName: string | null
  title: string | null
  suffix: string | null
  gender: string | null
  commissionClass: string | null

  nationality: string | null

  preferredLanguage: string | null

  // Channel / Hierarchy Codes
  channelCode: string | null
  subChannelCode: string | null

  agentLevel: string | null
  locationCode: string | null
  staffCode: string | null

  supervisor_Id: number | null
  contractedDate: string | null
  agentStatusCode: string | null
  statusDate: string | null

  isLicensed: boolean
  maskedPanNumber: string | null
  aadhaar_number: string | null
  irdaLicenseNumber: string | null
  gstNumber: string | null

  createdBy: string | null
  createdDate: string
  modifiedBy: string | null
  modifiedDate: string | null
  rowVersion: number | null
  isActive: boolean

  email: string | null

  supervisors?: IAgent[] | null
  reportees?: IAgent[] | null

  agentAuditTrail?: any[]
  peopleHeirarchy?: any[]
  total_count?: number

  father_Husband_Nm?: string | null

  employeeCode?: string | null
  applicationDocketNo?: string | null
  candidateType?: string | null
  startDate?: string | null
  appointmentDate?: string | null
  incorporationDate?: string | null

  agentTypeCategory?: string | null
  agentClassification?: string | null
  cmsAgentType?: string | null

  channel?: string | null
  sub_Channel?: string | null

  // Flags & Metadata
  ulipFlag: boolean
  trainingGroupType: string
  ifs: string
  refresherTrainingCompleted: boolean
  isMigrated: boolean
  mainPartnerClientCode: string
  agentMaincodevwEid: string
  registrationDate: string
  vertical: string

  branchCode: string

  // Trainings
  ic36TrngCompletionDate: string
  sTrngCompletionDate: string
  confirmationDate: string
  fgRockstarTrainingDate: string
  incrementDate: string
  lastPromotionDate: string
  hrDoj: string
  fgValueTrngDate: string
  hSecPolicyTrngDate: string
  itSecPolicyTrngDate: string
  npsTrngCompletionDate: string
  whistleBlowerTrngDate: string
  govPolicyTrngDate: string
  inductionTrngDate: string
  lastWorkingDate: string

  // Licensing
  licenseNo: string
  licenseType: string
  licenseIssueDate: string
  licenseExpiryDate: string
  licenseStatus: string | boolean

  // Contact Person
  cnctPersonDesig?: string
  cnctPersonMobileNo?: string
  cnctPersonEmail?: string
  cnctPersonName?: string

  // Backend bankAccounts (raw)
  bankAccounts?: IBankAccount[]

  // ===== Additional UI Fields You Added Manually =====
  pob?: string | null
  aadharNumber?: string | null
  educationQualification?: string | null
  educationSpecialization?: string | null
  educationalInstitute?: string | null
  passingYear?: string | null
  criminalRecord?: string | null
  employmentType?: string | null
  employmentStatus?: string | null
  experienceYears?: string | null
  uanNumber?: string | null
  reportingToName?: string | null
  reportingToDesignation?: string | null

  channelRegion?: string | null
  cluster?: string | null
  baseLocation?: string | null
  zone?: string | null
  irdaTrainingOrganization?: string | null

  rmName?: string | null
  rmMobile?: string | null
  rmEmail?: string | null
  smName?: string | null
  smMobile?: string | null
  smEmail?: string | null
  asmName?: string | null
  asmMobile?: string | null
  asmEmail?: string | null
  branchManagerMobile?: string | null
  branchManagerEmail?: string | null

  religion?: string | null
  caste?: string | null
  physicallyChallenged?: string | null

  birthIdentificationMark?: string | null

  workContactNo?: string | null
  residenceContactNo?: string | null

  // Financial
  panAadharLinkFlag?: boolean
  sec206abFlag?: boolean
  taxStatus: string
  serviceTaxNo: string
  // keyValueEntry: IkeyValueEntry[]
  
  //Acct Activation Date missing
  factoringHouse: string
  accountHolderName: string
  panNumber?: string
  bankName?: string
  branchName: string
  accountNumber: string
  accountType: string
  micr: string
  ifsc: string
  preferredPaymentMode: string

  // heirarchy

  designationCode: string | null
  designation: string | null

  //Other personal details
  dob: string | null
  martialStatus: string | null
  educationLevel: string | null
  education: string | null
  workProfile: string | null
  annualIncome: string | null
  // Nominees
  nominees?: IAgentNominees[]
  nomineeName?: string | null
  relationship?: string | null
  nomineeAge?: number | null
  occupation?: string | null
  urn?: string | null
  additionalComment?: string | null
  // Personal Info (backend nested)
  personalInfo?: IPersonalInfo[]
  workExpMonths: number
  bloodGroup: string
  birthPlace: string
  dateOfBirth: string

  // address
  permanentAddres: IpermanentAddres[]
  stateEid?: string | null
  addressType: string | null
  addressLine1?: boolean | null
  addressLine2?: boolean | null
  addressLine3?: boolean | null
  city?: string | null
  country?: string | null
  pin?: string | null
  state?: string | null

  bankAccType?: IkeyValueEntry[]
titles?: IkeyValueEntry[]
genders?: IkeyValueEntry[]
channelNames?: IkeyValueEntry[]
subChannels?: IkeyValueEntry[]
occupations?: IkeyValueEntry[]
agentTypeCategories?: IkeyValueEntry[]
agentClassifications?: IkeyValueEntry[]
maritalStatuses?: IkeyValueEntry[]
stateNames?: IkeyValueEntry[]
countries?: IkeyValueEntry[]

}

export interface IpermanentAddres {
  addressType: string
  refKey: string
  refType: string
  addressLine1: string
  addressLine2: string
  addressLine3: string
  city: string
  state: string
  country: string
  pin: string
  landmark: string
}
export interface IBankAccount {
  accountHolderName: string
  accountNumber: string
  ifsc: string
  micr: string
  bankName: string
  branchName: string
  accountType: string
  activeSince: string
  factoringHouse: string
  preferredPaymentMode: string
}
export interface IkeyValueEntry {
  orgid: number
  entryCategory: string
  entryIdentity: number
  entryDesc: string
  entryParentId: number
  activeStatus: boolean
}
export interface IAgentNominees {
  nomineeID: string
  nomineeName: string
  relationship: string
  nomineeAge: number
}
export interface IPersonalInfo {
  dateOfBirth: string
  panNumber: string
  email: string
  mobileNo: string
  workContactNo: string
  residenceContactNo: string
  bloodGroup: string
  birthPlace: string
  martialStatus: string
  educationLevel: string
  education: string
  workProfile: string
  annualIncome: string
  workExpMonths: number
}
export interface IAgentAuditTrail {
  agentCode: number
  agentId: string
  changedOn: string
  filedName: string
  oldValue: string
  newValue: string
  modifiedBy: string
  modifiedDate: string
  changeDescription: string
}
export interface IPeopleHierarchy {
  agentId: number
  agentCode: string | null
  firstName: string | null
  middleName: string | null
  lastName: string | null
  supervisors?: IPeopleHierarchy | null
}

export interface IEditAgentPayload {
  id: number | string
  sectionName: string
  [key: string]: any
}